xarray/core/indexing.py

```diff
@@ -419,14 +423,21 @@ def __setitem__(self, key, value):
         self.array[key] = value
 
 
-def xarray_indexable(array):
+def as_indexable(array):
+    """
+    This function always returns a NDArrayIndexable subclass,
+    so that the vectorized indexing is always possible with the returned
+    object.
+    """
+    if isinstance(array, NDArrayIndexable):
+        return array
     if isinstance(array, np.ndarray):
         return NumpyIndexingAdapter(array)
     if isinstance(array, pd.Index):
         return PandasIndexAdapter(array)
     if isinstance(array, dask_array_type):
         return DaskIndexingAdapter(array)
-    return array
+    raise TypeError('Invalid array type: {}'.format(type(array)))
```

Comment:
The default assumption previously was that any array inside an `xarray.Variable` supported indexing. With this change, we only support indexing when objects are instances of `NDArrayIndexable`. It seems like it might make more sense to rather never allow creating `Variable` objects with this invalid data in the first place, e.g., put this check in `xarray.core.variable.as_compatible_data()` instead.